ffi/rust/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+use std::ffi::{CStr, CString};
+
+use thiserror::Error;
+
+#[allow(non_camel_case_types,non_upper_case_globals,improper_ctypes)]
+pub mod ffi_wrapper;
+
+#[derive(Debug, Clone)]
+pub struct Config {
+    pub host_address: String,
+    pub resolver_address: String,
+    pub user_did: String,
+    pub agent_did: String,
+    pub agent_key_name: String,
+    pub agent_name: String,
+    pub agent_secret: String,
+    pub token_duration: i64,
+}
+
+#[derive(Error, Debug)]
+pub enum IdentityLibError {
+    #[error("FFI library error message: `{0}`.")]
+    Message(String),
+}
+
+impl Drop for ffi_wrapper::CreateAgentAuthToken_return {
+    fn drop(&mut self) {
+        unsafe {
+            ffi_wrapper::FreeUpCString(self.r0);
+            ffi_wrapper::FreeUpCString(self.r1);
+        }
+    }
+}
+
+impl Drop for ffi_wrapper::CreateTwinWithControlDelegation_return {
+    fn drop(&mut self) {
+        unsafe {
+            ffi_wrapper::FreeUpCString(self.r0);
+            ffi_wrapper::FreeUpCString(self.r1);
+            ffi_wrapper::FreeUpCString(self.r2);
+        }
+    }
+}
+
+pub fn create_agent_auth_token(options: &Config) -> Result<String, IdentityLibError> {
+    let agent_did_string = CString::new(options.agent_did.as_str()).expect("CString::new failed");
+    let agent_key_name_string = CString::new(options.agent_key_name.as_str()).expect("CString::new failed");
+    let agent_name_string = CString::new(options.agent_name.as_str()).expect("CString::new failed");
+    let user_did_string = CString::new(options.user_did.as_str()).expect("CString::new failed");
+    let seed_string = CString::new(options.agent_secret.as_str()).expect("CString::new failed");
```

Comment:
Instead of panicking with `expect`, should we gracefully return the error with `?`?